lib/veritech-server/src/server/pause_resume_stream.rs

```diff
@@ -0,0 +1,248 @@
+// TODO(nick,fletcher): make this generic and in its own crate.
+
+use std::{task::Poll, time::Duration};
+
+use futures::{future::BoxFuture, FutureExt, StreamExt};
+use pin_project_lite::pin_project;
+use si_data_nats::async_nats::jetstream::{
+    consumer::{self, Consumer, StreamError},
+    Message,
+};
+use telemetry::prelude::*;
+use tokio::sync::mpsc;
+
+#[remain::sorted]
+#[derive(Debug, Clone, Copy)]
+enum PauseResumeCommand {
+    // NOTE(nick,fletcer): consider adding "PauseWithDuration".
+    Pause,
+    Resume,
+}
+
+/// A controller for pausing and resuming a [`PauseResumeStream`].
+#[derive(Debug, Clone)]
+pub struct PauseResumeController {
+    inner: mpsc::UnboundedSender<PauseResumeCommand>,
+}
+
+impl PauseResumeController {
+    /// Pauses a [`PauseResumeStream`] with a stream-side timer.
+    pub fn pause(&self) {
+        // TODO(nick,fletcher): disambiguate the difference between "I failed to send due to an
+        // unexpected failure" vs. "I failed to send because we are shutting down and the rx has
+        // already closed for valid reasons".
+        if let Err(err) = self.inner.send(PauseResumeCommand::Pause) {
+            error!(si.error.message = ?err, "error when sending pause command");
+        }
+    }
+
+    /// Resumes a [`PauseResumeStream`].
+    #[allow(dead_code)]
+    pub fn resume(&self) {
+        // TODO(nick,fletcher): disambiguate the difference between "I failed to send due to an
+        // unexpected failure" vs. "I failed to send because we are shutting down and the rx has
+        // already closed for valid reasons".
+        if let Err(err) = self.inner.send(PauseResumeCommand::Resume) {
+            error!(si.error.message = ?err, "error when sending resume command");
+        }
+    }
+}
+
+#[remain::sorted]
+enum State {
+    ReadyToSubscribe,
+    Subscribed(consumer::pull::Stream),
+    Subscribing(BoxFuture<'static, Result<consumer::pull::Stream, StreamError>>),
+    Unsubscribed(BoxFuture<'static, ()>),
+}
+
+pin_project! {
```

Comment:
Used to mutate the fields. You can "project" the fields and then mutate them accordingly.